core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp

```diff
@@ -67,26 +67,102 @@ int hip_get_max_block_size(typename DriverType::functor_type const &f,
       f, vector_length, shmem_extra_block, shmem_extra_thread);
 }
 
-template <class FunctorType, class LaunchBounds>
-int hip_get_max_block_size(const HIPInternal * /*hip_instance*/,
-                           const hipFuncAttributes &attr,
-                           const FunctorType & /*f*/,
-                           const size_t /*vector_length*/,
-                           const size_t /*shmem_block*/,
-                           const size_t /*shmem_thread*/) {
-  // FIXME_HIP find a better algorithm. Be aware that
-  // maxThreadsPerMultiProcessor, regsPerBlock, and l2CacheSize are bugged and
-  // always return zero
-  // https://github.com/ROCm-Developer-Tools/HIP/blob/6c5fa32815650cc20a4f783d09b013610348a4d5/include/hip/hcc_detail/hip_runtime_api.h#L438-L440
-  // and we don't have access to the same information than we do for CUDA
-
-  int const max_threads_per_block_mi60 = 1024;
-  int const max_threads_per_block      = LaunchBounds::maxTperB == 0
-                                        ? max_threads_per_block_mi60
+template <class FunctorType, class LaunchBounds, typename F>
+int hip_internal_get_block_size(const F &condition_check,
+                                const HIPInternal *hip_instance,
+                                const hipFuncAttributes &attr,
+                                const FunctorType &f,
+                                const size_t vector_length,
+                                const size_t shmem_block,
+                                const size_t shmem_thread) {
+  const int min_blocks_per_sm =
+      LaunchBounds::minBperSM == 0 ? 1 : LaunchBounds::minBperSM;
+  const int max_threads_per_block = LaunchBounds::maxTperB == 0
+                                        ? hip_instance->m_maxThreadsPerBlock
                                         : LaunchBounds::maxTperB;
-  return std::min(attr.maxThreadsPerBlock, max_threads_per_block);
+
+  const int regs_per_wavefront  = attr.numRegs;
+  const int regs_per_sm         = hip_instance->m_regsPerSM;
+  const int shmem_per_sm        = hip_instance->m_shmemPerSM;
+  const int max_shmem_per_block = hip_instance->m_maxShmemPerBlock;
+  const int max_blocks_per_sm   = hip_instance->m_maxBlocksPerSM;
+  const int max_threads_per_sm  = hip_instance->m_maxThreadsPerSM;
+
+// FIXME_HIP this is broken in 3.5, but should be in 3.6
+#if (HIP_VERSION_MAJOR > 3 || HIP_VERSION_MINOR > 5 || \
+     HIP_VERSION_PATCH >= 20226)
```

Comment:
ummm.... this will be true for HIP 2.7 or whatever (for instance; I know that probably doesn't exist) or 3.4 with version patch 31415. I doubt that's the intention.

Reply:
This is just temporary. We only support the latest version of the compiler so HIP 2.7 or 3.4 won't work anyway. As soon as HIP 3.6 is released, we will require it and this can be removed.